components/locale/src/langid.rs

```diff
@@ -0,0 +1,151 @@
+use std::fmt::Write;
+use std::str::FromStr;
+
+use crate::parser::{parse_language_identifier, ParserError, ParserMode};
+use crate::subtags;
+
+/// `LanguageIdentifier` is a core struct representing a [`Unicode Language Identifier`].
+///
+/// # Example
+///
+/// ```
+/// use icu_locale::LanguageIdentifier;
+///
+/// let li: LanguageIdentifier = "en-US".parse()
+///     .expect("Failed to parse.");
+///
+/// assert_eq!(li.language, "en");
+/// assert_eq!(li.script, None);
+/// assert_eq!(li.region.unwrap(), "US");
+/// assert_eq!(li.variants.len(), 0);
+/// ```
+///
+/// # Parsing
+///
+/// Unicode recognizes three levels of standard conformance for any language identifier:
+///
+///  * *well-formed* - syntactically correct
+///  * *valid* - well-formed and only uses registered language subtags, extensions, keywords, types...
+///  * *canonical* - valid and no deprecated codes or structure.
+///
+/// At the moment parsing normalizes a well-formed language identifier converting
+/// `_` separators to `-` and adjusting casing to conform to the Unicode standard.
+///
+/// Any bogus subtags will cause the parsing to fail with an error.
+/// No subtag validation is performed.
+///
+/// # Example:
+///
+/// ```
+/// use icu_locale::LanguageIdentifier;
+///
+/// let li: LanguageIdentifier = "eN_latn_Us-Valencia".parse()
+///     .expect("Failed to parse.");
+///
+/// assert_eq!(li.language, "en");
+/// assert_eq!(li.script.unwrap(), "Latn");
+/// assert_eq!(li.region.unwrap(), "US");
+/// assert_eq!(li.variants.get(0).unwrap(), "valencia");
+/// ```
+///
+/// [`Unicode Language Identifier`]: https://unicode.org/reports/tr35/tr35.html#Unicode_language_identifier
+#[derive(Default, Debug, PartialEq, Eq, Clone, Hash, PartialOrd, Ord)]
+pub struct LanguageIdentifier {
```

Comment:
(Not sure if this was already resolved in another thread.)
Is the plan to make these fields private and add getters in a future PR? Strongly committing to a specific internal structure in such a commonly used struct seems like a risky move.

Reply:
No, the plan is to keep them open.

We don't strongly commit to specific internal structure, we provide public abstraction for each of the fields - in other words, no user will ever deal with anything but `Language`, `Region`, `Script`, `Variant` structs and `LanguageIdentifier` is just a collection of those.

Same thing happens with all extensions in the follow up PR.

Reply:
I'm going to resolve this. If you're concerned about it still, please, file an issue and let's discuss it there.